src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs

```diff
@@ -49,29 +49,27 @@ public override int Run(InterpretedFrame frame)
         }
     }
 
-    internal sealed class NewInstruction : Instruction
+    internal class NewInstruction : Instruction
     {
-        private readonly ConstructorInfo _constructor;
-        private readonly int _argCount;
+        protected readonly ConstructorInfo _constructor;
+        protected readonly int _argumentCount;
 
-        public NewInstruction(ConstructorInfo constructor)
+        public NewInstruction(ConstructorInfo constructor, int argumentCount)
         {
             _constructor = constructor;
-            _argCount = constructor.GetParameters().Length;
+            _argumentCount = argumentCount;
         }
-        public override int ConsumedStack { get { return _argCount; } }
+        public override int ConsumedStack { get { return _argumentCount; } }
         public override int ProducedStack { get { return 1; } }
         public override string InstructionName
         {
             get { return "New"; }
         }
         public override int Run(InterpretedFrame frame)
         {
-            object[] args = new object[_argCount];
-            for (int i = _argCount - 1; i >= 0; i--)
-            {
-                args[i] = frame.Pop();
```

Comment:
Made analogous to `CallInstruction`; there's no need to do a bunch of `Pop`s which decrement the `StackIndex` if we can just obtain the arguments from the frame. (Note that `Pop` doesn't clear frame slots; it just reads and adjusts the pointer).